lib/mayaUsd/ufe/UsdTRSUndoableCommandBase.cpp

```diff
@@ -0,0 +1,112 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "UsdTRSUndoableCommandBase.h"
+#include "private/Utils.h"
+
+#include <ufe/scene.h>
+#include <ufe/sceneNotification.h>
+
+MAYAUSD_NS_DEF {
+namespace ufe {
+
+template<class V>
+UsdTRSUndoableCommandBase<V>::UsdTRSUndoableCommandBase(
+    const UsdSceneItem::Ptr& item, double x, double y, double z
+) : fItem(item), fNewValue(x, y, z)
+{}
+
+template<class V>
+void UsdTRSUndoableCommandBase<V>::initialize()
+{
+    if (cannotInit()) {
+        return;
+    }
+
+    // If prim does not have the attribute, add it.
+    if (!prim().HasAttribute(attributeName()))
+    {
+        fOpAdded = true;
+        addEmptyAttribute();
+    }
+
+    // See
+    // https://stackoverflow.com/questions/17853212/using-shared-from-this-in-templated-classes
+    // for explanation of this->shared_from_this() in templated class.
+    attribute().Get(&fPrevValue);
+    Ufe::Scene::instance().addObjectPathChangeObserver(this->shared_from_this());
+}
+
+template<class V>
+void UsdTRSUndoableCommandBase<V>::operator()(
+    const Ufe::Notification& n
+)
+{
+    if (auto renamed = dynamic_cast<const Ufe::ObjectRename*>(&n)) {
+        checkNotification(renamed);
+    }
+    else if (auto reparented = dynamic_cast<const Ufe::ObjectReparent*>(&n)) {
+        checkNotification(reparented);
+    }
+}
+
+template<class V>
+void UsdTRSUndoableCommandBase<V>::undoImp()
+{
+    attribute().Set(fPrevValue);
+    // Todo : We would want to remove the xformOp
+    // (SD-06/07/2018) Haven't found a clean way to do it - would need to investigate
+}
+
+template<class V>
+void UsdTRSUndoableCommandBase<V>::redoImp()
+{
+    if (fDoneOnce) {
+        attribute().Set(fNewValue);
```

Comment:
If the command has been done once, we know there is an attribute on the prim for the translate, rotate, or scale.  Otherwise, we must invoke the code that creates the attribute if required.

Reply:
The attribute is added, but perform() does conversion to the common transform API.  Without it, assertions occur on rotate and scale, such as this one:
Runtime Error: in _SetValueImpl at line 5604 of /mnt/dataTwoTwenty/tremblp/projects/ecg-usd-build/ecg-usd-build/usd/pxr/usd/usd/stage.cpp -- Empty typeName for </pCube1.xformOp:rotateXYZ>
I will add this comment to the implementation.